test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +102,392 @@ UCS_TEST_P(test_ucp_request, test_request_query)
+        wait_for_value(&m_completed, (unsigned)1);
+        ucp_request_release(request);
+    }
+
+    void send_nb(std::vector<char> &sbuf, std::vector<char> *rbuf_p,
+                 operation_e op, bool sync, std::vector<void*> &reqs)
+    {
+        ucp_request_param_t param = {0};
+        void *rreq                = NULL;
+        void *sreq                = NULL;
+        std::vector<char> &rbuf   = *rbuf_p;
+        sbuf.resize(m_msg_size);
+        rbuf.resize(m_msg_size);
+        ucs::fill_random(sbuf);
+        ucs::fill_random(rbuf);
+
+        switch (op) {
+        case TAG:
+            param.op_attr_mask = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+            if (sync) {
+                sreq = ucp_tag_send_sync_nbx(sender().ep(), sbuf.data(),
+                                             m_msg_size, 0, &param);
+            } else {
+                sreq = ucp_tag_send_nbx(sender().ep(), sbuf.data(), m_msg_size,
+                                        0, &param);
+            }
+
+            ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+            rreq = ucp_tag_recv_nbx(receiver().worker(), rbuf.data(),
+                                    m_msg_size, 0, 0, &param);
+            ASSERT_TRUE(UCS_PTR_IS_PTR(rreq));
+            reqs.push_back(rreq);
+            break;
+
+        case RMA_GET:
+            init_rkey(sbuf);
+            param.op_attr_mask = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+            sreq = ucp_get_nbx(sender().ep(), rbuf.data(), m_msg_size,
+                               (uint64_t)sbuf.data(), m_rkeys.back(), &param);
+            break;
+
+        case RMA_PUT:
+            init_rkey(sbuf);
+            sreq = ucp_put_nbx(sender().ep(), rbuf.data(), m_msg_size,
+                               (uint64_t)sbuf.data(), m_rkeys.back(), &param);
+            break;
+
+        case STREAM:
+            param.op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE;
+            param.datatype     = ucp_dt_make_contig(1);
+            sreq = ucp_stream_send_nbx(sender().ep(), sbuf.data(), m_msg_size,
+                                       &param);
+            break;
+
+        case AM:
+            sreq = send_am(sbuf, rbuf_p);
+            break;
+        }
+
+        if (sreq != NULL) {
+            ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+        }
+
+        reqs.push_back(sreq);
+    }
+
+    void reset_protocol(operation_e op, bool sync = false)
+    {
+        static const unsigned reqs_count = 1000;
+        std::vector<void*> reqs;
+        m_sbufs.resize(reqs_count);
+        m_rbufs.resize(reqs_count);
+
+        for (int i = 0; i < reqs_count; ++i) {
+            send_nb(m_sbufs[i], &m_rbufs[i], op, sync, reqs);
+        }
+
+        restart(sender().ep());
+
+        for (int i = 0; i < reqs_count; ++i) {
+            if (op == STREAM) {
+                get_stream_data(m_rbufs[i]);
+            } else if (op == AM) {
+                wait_for_condition(
+                        [this, i]() { return m_rbufs[i] == m_sbufs[i]; });
```

Comment:
maybe better count active message callbacks count? and then validate the data in line 349